File: packages/paygate-node/src/cli/audit.ts
import { readFileSync } from 'node:fs';
import { AuditLogger } from '../analytics/audit-log.js';

export async function runAudit(action: string, opts: { file?: string }): Promise<void> {
  if (!opts.file) throw new Error('--file <path> is required');
  switch (action) {
    case 'verify': {
      const result = AuditLogger.verify(opts.file);
      if (result.ok) console.log(`OK — ${result.rows} rows verified`);
      else {
        console.error(`BROKEN at row ${result.brokenAt}`);
        process.exitCode = 1;
      }
      return;
    }
    case 'tail': {
      const raw = readFileSync(opts.file, 'utf-8');
      const lines = raw.trimEnd().split('\n');
      for (const line of lines.slice(-20)) console.log(line);
      return;
    }
    case 'pack': {
      console.error('audit pack: implemented in paygate-core (out of scope for the node CLI)');
      process.exitCode = 1;
      return;
    }
    default:
      console.error(`unknown action: ${action}`);
      process.exitCode = 1;
  }
}

File: packages/paygate-python/paygate/cli/audit.py
"""`paygate audit` — verify / tail the hash-chained audit log."""

from __future__ import annotations

import pathlib

import typer

from paygate.analytics.audit_log import AuditLogger

audit_app = typer.Typer(no_args_is_help=True)


@audit_app.command("verify")
def verify(file: str = typer.Option(..., "--file")) -> None:
    result = AuditLogger.verify(pathlib.Path(file))
    if result.ok:
        typer.echo(f"OK — {result.rows} rows verified")
    else:
        typer.echo(f"BROKEN at row {result.broken_at}", err=True)
        raise typer.Exit(code=1)


@audit_app.command("tail")
def tail(
    file: str = typer.Option(..., "--file"),
    n: int = typer.Option(20, "-n", help="last N lines"),
) -> None:
    raw = pathlib.Path(file).read_text("utf-8").splitlines()
    for line in raw[-n:]:
        typer.echo(line)
